// yerror/dist/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = exports.YError = void 0;

var _os = _interopRequireDefault(require("os"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } }

function _createClass(Constructor, protoProps, staticProps) { if (protoProps) _defineProperties(Constructor.prototype, protoProps); if (staticProps) _defineProperties(Constructor, staticProps); Object.defineProperty(Constructor, "prototype", { writable: false }); return Constructor; }

function _possibleConstructorReturn(self, call) { if (call && (typeof call === "object" || typeof call === "function")) { return call; } else if (call !== void 0) { throw new TypeError("Derived constructors may only return object or undefined"); } return _assertThisInitialized(self); }

function _assertThisInitialized(self) { if (self === void 0) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function"); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, writable: true, configurable: true } }); Object.defineProperty(subClass, "prototype", { writable: false }); if (superClass) _setPrototypeOf(subClass, superClass); }

function _wrapNativeSuper(Class) { var _cache = typeof Map === "function" ? new Map() : undefined; _wrapNativeSuper = function _wrapNativeSuper(Class) { if (Class === null || !_isNativeFunction(Class)) return Class; if (typeof Class !== "function") { throw new TypeError("Super expression must either be null or a function"); } if (typeof _cache !== "undefined") { if (_cache.has(Class)) return _cache.get(Class); _cache.set(Class, Wrapper); } function Wrapper() { return _construct(Class, arguments, _getPrototypeOf(this).constructor); } Wrapper.prototype = Object.create(Class.prototype, { constructor: { value: Wrapper, enumerable: false, writable: true, configurable: true } }); return _setPrototypeOf(Wrapper, Class); }; return _wrapNativeSuper(Class); }

function _construct(Parent, args, Class) { if (_isNativeReflectConstruct()) { _construct = Reflect.construct.bind(); } else { _construct = function _construct(Parent, args, Class) { var a = [null]; a.push.apply(a, args); var Constructor = Function.bind.apply(Parent, a); var instance = new Constructor(); if (Class) _setPrototypeOf(instance, Class.prototype); return instance; }; } return _construct.apply(null, arguments); }

function _isNativeReflectConstruct() { if (typeof Reflect === "undefined" || !Reflect.construct) return false; if (Reflect.construct.sham) return false; if (typeof Proxy === "function") return true; try { Boolean.prototype.valueOf.call(Reflect.construct(Boolean, [], function () {})); return true; } catch (e) { return false; } }

function _isNativeFunction(fn) { return Function.toString.call(fn).indexOf("[native code]") !== -1; }

function _setPrototypeOf(o, p) { _setPrototypeOf = Object.setPrototypeOf ? Object.setPrototypeOf.bind() : function _setPrototypeOf(o, p) { o.__proto__ = p; return o; }; return _setPrototypeOf(o, p); }

function _getPrototypeOf(o) { _getPrototypeOf = Object.setPrototypeOf ? Object.getPrototypeOf.bind() : function _getPrototypeOf(o) { return o.__proto__ || Object.getPrototypeOf(o); }; return _getPrototypeOf(o); }

/**
 * An YError class able to contain some params and
 *  print better stack traces
 * @extends Error
 */
let YError = /*#__PURE__*/function (_Error) {
  _inherits(YError, _Error);

  /**
   * Creates a new YError with an error code
   *  and some params as debug values.
   * @param {string} [errorCode = 'E_UNEXPECTED']
   * The error code corresponding to the actual error
   * @param {...any} [params]
   * Some additional debugging values
   */
  function YError(wrappedErrors, errorCode, ...params) {
    var _this;

    _classCallCheck(this, YError);

    // Detecting if wrappedErrors are passed
    if (!(wrappedErrors instanceof Array)) {
      params = ('undefined' === typeof errorCode ? [] : [errorCode]).concat(params);
      errorCode = wrappedErrors;
      wrappedErrors = [];
    } // Call the parent constructor


    _this = _possibleConstructorReturn(this, _getPrototypeOf(YError).call(this, errorCode)); // Filling error

    _this.code = errorCode || 'E_UNEXPECTED';
    _this.params = params;
    _this.wrappedErrors = wrappedErrors;
    _this.name = _this.toString();

    if (Error.captureStackTrace) {
      Error.captureStackTrace(_assertThisInitialized(_this), _this.constructor);
    }

    return _this;
  }

  _createClass(YError, [{
    key: "toString",
    value: function toString() {
      return (this.wrappedErrors.length ? // eslint-disable-next-line
      this.wrappedErrors[this.wrappedErrors.length - 1].stack + _os.default.EOL : '') + this.constructor.name + ': ' + this.code + ' (' + this.params.join(', ') + ')';
    }
  }]);

  return YError;
}(_wrapNativeSuper(Error));
/**
 * Wraps any error and output a YError with an error
 *  code and some params as debug values.
 * @param {Error} err
 * The error to wrap
 * @param {string} [errorCode = 'E_UNEXPECTED']
 * The error code corresponding to the actual error
 * @param {...any} [params]
 * Some additional debugging values
 * @return {YError}
 * The wrapped error
 */


exports.YError = YError;

YError.wrap = function yerrorWrap(err, errorCode, ...params) {
  let yError = null;

  const wrappedErrorIsACode = _looksLikeAYErrorCode(err.message);

  const wrappedErrors = (err.wrappedErrors || []).concat(err);

  if (!errorCode) {
    if (wrappedErrorIsACode) {
      errorCode = err.message;
    } else {
      errorCode = 'E_UNEXPECTED';
    }
  }

  if (err.message && !wrappedErrorIsACode) {
    params.push(err.message);
  }

  yError = new YError(wrappedErrors, errorCode, ...params);
  return yError;
};
/**
 * Return a YError as is or wraps any other error and output
 *  a YError with a code and some params as debug values.
 * @param {Error} err
 * The error to cast
 * @param {string} [errorCode = 'E_UNEXPECTED']
 * The error code corresponding to the actual error
 * @param {...any} [params]
 * Some additional debugging values
 * @return {YError}
 * The wrapped error
 */


YError.cast = function yerrorCast(err, ...params) {
  if (_looksLikeAYError(err)) {
    return err;
  }

  return YError.wrap.apply(YError, [err].concat(params));
};
/**
 * Same than `YError.wrap()` but preserves the code
 *  and the debug values of the error if it is
 *  already an instance of the YError constructor.
 * @param {Error} err
 * The error to bump
 * @param {string} [errorCode = 'E_UNEXPECTED']
 * The error code corresponding to the actual error
 * @param {...any} [params]
 * Some additional debugging values
 * @return {YError}
 * The wrapped error
 */


YError.bump = function yerrorBump(err, ...params) {
  if (_looksLikeAYError(err)) {
    return YError.wrap.apply(YError, [err, err.code].concat(err.params));
  }

  return YError.wrap.apply(YError, [err].concat(params));
}; // In order to keep compatibility through major versions
// we have to make kind of an cross major version instanceof


function _looksLikeAYError(err) {
  return err instanceof YError || err.constructor && err.constructor.name && err.constructor.name.endsWith('Error') && 'string' === typeof err.code && _looksLikeAYErrorCode(err.code) && err.params && err.params instanceof Array;
}

function _looksLikeAYErrorCode(str) {
  return /^([A-Z0-9_]+)$/.test(str);
}

var _default = YError;
exports.default = _default;
